### ISSUE TITLE
Cannot refresh lxd, systemd files refuse to stop

### ISSUE THREAD

**Issue by bakerjalexander** [lxc/lxd#7500]
Distribution: Ubuntu
Distribution version: Ubuntu 18.04.4 LTS
   lxc info:
config: {}
api_extensions:
- storage_zfs_remove_snapshots
- container_host_shutdown_timeout
- container_stop_priority
- container_syscall_filtering
- auth_pki
- container_last_used_at
- etag
- patch
- usb_devices
- https_allowed_credentials
- image_compression_algorithm
- directory_manipulation
- container_cpu_time
- storage_zfs_use_refquota
- storage_lvm_mount_options
- network
- profile_usedby
- container_push
- container_exec_recording
- certificate_update
- container_exec_signal_handling
- gpu_devices
- container_image_properties
- migration_progress
- id_map
- network_firewall_filtering
- network_routes
- storage
- file_delete
- file_append
- network_dhcp_expiry
- storage_lvm_vg_rename
- storage_lvm_thinpool_rename
- network_vlan
- image_create_aliases
- container_stateless_copy
- container_only_migration
- storage_zfs_clone_copy
- unix_device_rename
- storage_lvm_use_thinpool
- storage_rsync_bwlimit
- network_vxlan_interface
- storage_btrfs_mount_options
- entity_description
- image_force_refresh
- storage_lvm_lv_resizing
- id_map_base
- file_symlinks
- container_push_target
- network_vlan_physical
- storage_images_delete
- container_edit_metadata
- container_snapshot_stateful_migration
- storage_driver_ceph
- storage_ceph_user_name
- resource_limits
- storage_volatile_initial_source
- storage_ceph_force_osd_reuse
- storage_block_filesystem_btrfs
- resources
- kernel_limits
- storage_api_volume_rename
- macaroon_authentication
- network_sriov
- console
- restrict_devlxd
- migration_pre_copy
- infiniband
- maas_network
- devlxd_events
- proxy
- network_dhcp_gateway
- file_get_symlink
- network_leases
- unix_device_hotplug
- storage_api_local_volume_handling
- operation_description
- clustering
- event_lifecycle
- storage_api_remote_volume_handling
- nvidia_runtime
- container_mount_propagation
- container_backup
- devlxd_images
- container_local_cross_pool_handling
- proxy_unix
- proxy_udp
- clustering_join
- proxy_tcp_udp_multi_port_handling
- network_state
- proxy_unix_dac_properties
- container_protection_delete
- unix_priv_drop
- pprof_http
- proxy_haproxy_protocol
- network_hwaddr
- proxy_nat
- network_nat_order
- container_full
- candid_authentication
- backup_compression
- candid_config
- nvidia_runtime_config
- storage_api_volume_snapshots
- storage_unmapped
- projects
- candid_config_key
- network_vxlan_ttl
- container_incremental_copy
- usb_optional_vendorid
- snapshot_scheduling
- container_copy_project
- clustering_server_address
- clustering_image_replication
- container_protection_shift
- snapshot_expiry
- container_backup_override_pool
- snapshot_expiry_creation
- network_leases_location
- resources_cpu_socket
- resources_gpu
- resources_numa
- kernel_features
- id_map_current
- event_location
- storage_api_remote_volume_snapshots
- network_nat_address
- container_nic_routes
- rbac
- cluster_internal_copy
- seccomp_notify
- lxc_features
- container_nic_ipvlan
- network_vlan_sriov
- storage_cephfs
- container_nic_ipfilter
- resources_v2
- container_exec_user_group_cwd
- container_syscall_intercept
- container_disk_shift
- storage_shifted
- resources_infiniband
- daemon_storage
- instances
- image_types
- resources_disk_sata
- clustering_roles
- images_expiry
- resources_network_firmware
- backup_compression_algorithm
- ceph_data_pool_name
- container_syscall_intercept_mount
- compression_squashfs
- container_raw_mount
- container_nic_routed
- container_syscall_intercept_mount_fuse
- container_disk_ceph
- virtual-machines
- image_profiles
- clustering_architecture
- resources_disk_id
- storage_lvm_stripes
- vm_boot_priority
- unix_hotplug_devices
- api_filtering
- instance_nic_network
- clustering_sizing
- firewall_driver
- projects_limits
- container_syscall_intercept_hugetlbfs
- limits_hugepages
- container_nic_routed_gateway
- projects_restrictions
- custom_volume_snapshot_expiry
- volume_snapshot_scheduling
- trust_ca_certificates
- snapshot_disk_usage
- clustering_edit_roles
- container_nic_routed_host_address
- container_nic_ipvlan_gateway
- resources_usb_pci
- resources_cpu_threads_numa
- resources_cpu_core_die
- api_os
- container_nic_routed_host_table
- container_nic_ipvlan_host_table
- container_nic_ipvlan_mode
- resources_system
- images_push_relay
api_status: stable
api_version: "1.0"
auth: trusted
public: false
auth_methods:
- tls
environment:
  addresses: []
  architectures:
  - x86_64
  - i686
  certificate: xxxxxxxxxxxxxxxxxxxxxxxxxxxxx
  certificate_fingerprint: xxxxxxxxxxxxxxxxxxx
  driver: lxc
  driver_version: 4.0.2
  firewall: xtables
  kernel: Linux
  kernel_architecture: x86_64
  kernel_features:
    netnsid_getifaddrs: "false"
    seccomp_listener: "false"
    seccomp_listener_continue: "false"
    shiftfs: "false"
    uevent_injection: "false"
    unpriv_fscaps: "true"
  kernel_version: 4.15.0-99-generic
  lxc_features:
    cgroup2: "true"
    mount_injection_file: "true"
    network_gateway_device_route: "true"
    network_ipvlan: "true"
    network_l2proxy: "true"
    network_phys_macvlan_mtu: "true"
    network_veth_router: "true"
    pidfd: "false"
    seccomp_notify: "true"
  os_name: Ubuntu
  os_version: "18.04"
  project: default
  server: lxd
  server_clustered: false
  server_name: my server
  server_pid: 4718
  server_version: "4.1"
  storage: dir | zfs
  storage_version: 1 | 0.8.3-1ubuntu6~18.04.york0
  
 Storage backend in use: zfs

Issue description

"snap refresh lxd" doesn't work. Instead, I get:

INFO Waiting for snap.lxd.daemon.unix.socket to stop.
(Repeat 20 times)
INFO Waiting for snap.lxd.daemon.service to stop.
(Repeat 20 times)
INFO snap.lxd.daemon.service refused to stop, killing.
2020-06-08T11:54:35-05:00 ERROR snap.lxd.daemon.unix.socket failed to stop:
timeout
error: cannot perform the following tasks:
- Mount snap "lxd" (15359) (snap-lxd-15359.mount failed to stop: timeout)
- Stop snap "lxd" services (snap.lxd.daemon.unix.socket failed to stop: timeout)

Also, I lose the unix socket and have to manually restart with "systemctl start snap.lxd.activate.service"

I've tried manually downloading the snap and mounting it, then copying lxd to lxd.debug
"cp /mnt/bin/lxd /var/snap/lxd/common/lxd.debug" and running lxd --debug --group lxd. The debug runs and nothing out of the ordinary shows up but "snap refresh lxd" still gives the same result.

# Steps to reproduce 

 1. run "snap refresh lxd"

debug output:

LXD 4.1 is starting in normal mode       path=/var/snap/lxd/common/lxd
INFO[06-08|12:13:33] Kernel uid/gid map: 
INFO[06-08|12:13:33]  - u 0 0 4294967295 
INFO[06-08|12:13:33]  - g 0 0 4294967295 
INFO[06-08|12:13:33] Configured LXD uid/gid map: 
INFO[06-08|12:13:33]  - u 0 1000000 1000000000 
INFO[06-08|12:13:33]  - g 0 1000000 1000000000 
INFO[06-08|12:13:33] Kernel features: 
INFO[06-08|12:13:33]  - netnsid-based network retrieval: no 
INFO[06-08|12:13:33]  - pidfds: no 
INFO[06-08|12:13:33]  - uevent injection: no 
INFO[06-08|12:13:33]  - seccomp listener: no 
INFO[06-08|12:13:33]  - seccomp listener continue syscalls: no 
INFO[06-08|12:13:33]  - unprivileged file capabilities: yes 
INFO[06-08|12:13:33]  - cgroup layout: hybrid 
WARN[06-08|12:13:33]  - Couldn't find the CGroup memory swap accounting, swap limits will be ignored 
INFO[06-08|12:13:33]  - shiftfs support: no 
INFO[06-08|12:13:33] Initializing local database 
DBUG[06-08|12:13:33] Initializing database gateway 
DBUG[06-08|12:13:33] Start database node                      id=1 address= role=voter
INFO[06-08|12:13:34] Starting /dev/lxd handler: 
INFO[06-08|12:13:34]  - binding devlxd socket                 socket=/var/snap/lxd/common/lxd/devlxd/sock
INFO[06-08|12:13:34] REST API daemon: 
INFO[06-08|12:13:34]  - binding Unix socket                   socket=/var/snap/lxd/common/lxd/unix.socket
INFO[06-08|12:13:34] Initializing global database 
DBUG[06-08|12:13:34] Dqlite: attempt 0: server 1: connected 
DBUG[06-08|12:13:35] Firewall detected "nftables" incompatibility: Kernel version does not meet minimum requirement of 5 
DBUG[06-08|12:13:35] Firewall xtables detected iptables is in use 
INFO[06-08|12:13:35] Firewall loaded driver "xtables" 
INFO[06-08|12:13:35] Initializing storage pools 
DBUG[06-08|12:13:35] Initializing and checking storage pool "lxd" 
DBUG[06-08|12:13:35] Mount started                            driver=dir pool=lxd
DBUG[06-08|12:13:35] Mount finished                           driver=dir pool=lxd
DBUG[06-08|12:13:35] Initializing and checking storage pool "lxcpool" 
DBUG[06-08|12:13:35] Mount started                            driver=zfs pool=lxcpool
DBUG[06-08|12:13:35] Mount finished                           driver=zfs pool=lxcpool
DBUG[06-08|12:13:35] Initializing and checking storage pool "rfast" 
DBUG[06-08|12:13:35] Mount started                            driver=zfs pool=rfast
DBUG[06-08|12:13:35] Mount finished                           driver=zfs pool=rfast
INFO[06-08|12:13:35] Initializing daemon storage mounts 
INFO[06-08|12:13:35] Initializing networks 
DBUG[06-08|12:13:35] New task Operation: a147a5c8-c828-4c58-94f0-2b29acc59cd4 
INFO[06-08|12:13:35] Pruning leftover image files 
DBUG[06-08|12:13:35] Started task operation: a147a5c8-c828-4c58-94f0-2b29acc59cd4 
INFO[06-08|12:13:35] Done pruning leftover image files 
INFO[06-08|12:13:35] Loading daemon configuration 
DBUG[06-08|12:13:35] Success for task operation: a147a5c8-c828-4c58-94f0-2b29acc59cd4 
DBUG[06-08|12:13:35] Initialized inotify with file descriptor 22 
DBUG[06-08|12:13:35] New task Operation: 905b923a-30f1-4a44-b767-01a72b432745 
INFO[06-08|12:13:35] Pruning expired images 
DBUG[06-08|12:13:35] Started task operation: 905b923a-30f1-4a44-b767-01a72b432745 
INFO[06-08|12:13:35] Done pruning expired images 
DBUG[06-08|12:13:35] New task Operation: e417d343-1d4e-4278-b005-4496ef0fcd12 
DBUG[06-08|12:13:35] Success for task operation: 905b923a-30f1-4a44-b767-01a72b432745 
INFO[06-08|12:13:35] Pruning expired instance backups 
DBUG[06-08|12:13:35] Started task operation: e417d343-1d4e-4278-b005-4496ef0fcd12 
INFO[06-08|12:13:35] Done pruning expired instance backups 
DBUG[06-08|12:13:35] Success for task operation: e417d343-1d4e-4278-b005-4496ef0fcd12 
DBUG[06-08|12:13:35] New task Operation: 6d39d630-8b67-40c8-8e5f-231405e8085c 
DBUG[06-08|12:13:35] New task Operation: 5f4b38f0-1a3a-4982-a87d-1115adda6c4c 
DBUG[06-08|12:13:35] New task Operation: 7b4bf1a8-461f-4310-b1e8-c7a73e83063d 
INFO[06-08|12:13:35] Updating images 
DBUG[06-08|12:13:35] Started task operation: 5f4b38f0-1a3a-4982-a87d-1115adda6c4c 
INFO[06-08|12:13:35] Updating instance types 
DBUG[06-08|12:13:35] Started task operation: 7b4bf1a8-461f-4310-b1e8-c7a73e83063d 
INFO[06-08|12:13:35] Expiring log files 
DBUG[06-08|12:13:35] Started task operation: 6d39d630-8b67-40c8-8e5f-231405e8085c 
INFO[06-08|12:13:35] Done updating images 
INFO[06-08|12:13:35] Done expiring log files 
INFO[06-08|12:13:35] Done updating instance types 
DBUG[06-08|12:13:35] Success for task operation: 6d39d630-8b67-40c8-8e5f-231405e8085c 
DBUG[06-08|12:13:35] Processing image                         fp=d125be3eded77d54fdeb2971e25b7f98b8e80a71f365353def505ac0c7c63803 server=https://images.linuxcontainers.org protocol=simplestreams alias=ubuntu/bionic/amd64
DBUG[06-08|12:13:35] Connecting to a remote simplestreams server 
DBUG[06-08|12:13:36] Scheduler: network: vethb1196dee has been added: updating network priorities 
DBUG[06-08|12:13:36] Scheduler: network: veth7cc89c56 has been added: updating network priorities 
DBUG[06-08|12:13:36] MountInstance started                    driver=zfs pool=lxcpool project=default instance=apache
DBUG[06-08|12:13:36] Mounted ZFS dataset                      driver=zfs pool=lxcpool dev=rpool/lxc/containers/apache path=/var/snap/lxd/common/lxd/storage-pools/lxcpool/containers/apache
DBUG[06-08|12:13:36] MountInstance finished                   driver=zfs pool=lxcpool project=default instance=apache
DBUG[06-08|12:13:36] UpdateInstanceBackupFile started         driver=zfs pool=lxcpool project=default instance=apache
DBUG[06-08|12:13:36] UpdateInstanceBackupFile finished        driver=zfs pool=lxcpool project=default instance=apache
DBUG[06-08|12:13:36] MountInstance started                    driver=zfs pool=lxcpool project=default instance=apache
DBUG[06-08|12:13:36] MountInstance finished                   driver=zfs pool=lxcpool project=default instance=apache
INFO[06-08|12:13:36] Starting container                       used=2020-06-08T11:01:38-0500 stateful=false project=default name=apache action=start created=2020-04-09T15:03:04-0500 ephemeral=false
DBUG[06-08|12:13:37] Handling                                 method=GET url=/internal/containers/55/onstart ip=@ user=
DBUG[06-08|12:13:37] MountInstance started                    driver=zfs pool=lxcpool project=default instance=apache
DBUG[06-08|12:13:37] MountInstance finished                   driver=zfs pool=lxcpool project=default instance=apache
DBUG[06-08|12:13:37] Scheduler: container apache started: re-balancing 
INFO[06-08|12:13:37] Started container                        action=start created=2020-04-09T15:03:04-0500 ephemeral=false used=2020-06-08T11:01:38-0500 stateful=false project=default name=apache
DBUG[06-08|12:13:37] MountCustomVolume started                driver=zfs pool=rfast project=default volName=mysql-data
DBUG[06-08|12:13:37] Mounted ZFS dataset                      driver=zfs pool=rfast dev=rfast/custom/default_mysql-data path=/var/snap/lxd/common/lxd/storage-pools/rfast/custom/default_mysql-data
DBUG[06-08|12:13:37] MountCustomVolume finished               driver=zfs pool=rfast project=default volName=mysql-data
DBUG[06-08|12:13:37] MountCustomVolume started                driver=zfs pool=rfast project=default volName=mysql-log
DBUG[06-08|12:13:37] Mounted ZFS dataset                      driver=zfs pool=rfast path=/var/snap/lxd/common/lxd/storage-pools/rfast/custom/default_mysql-log dev=rfast/custom/default_mysql-log
DBUG[06-08|12:13:37] MountCustomVolume finished               driver=zfs pool=rfast project=default volName=mysql-log
DBUG[06-08|12:13:37] Scheduler: network: veth1f6bc0bd has been added: updating network priorities 
DBUG[06-08|12:13:37] Scheduler: network: vethfa763dd8 has been added: updating network priorities 
DBUG[06-08|12:13:37] MountInstance started                    pool=rfast driver=zfs project=default instance=mysql
DBUG[06-08|12:13:37] Mounted ZFS dataset                      pool=rfast driver=zfs path=/var/snap/lxd/common/lxd/storage-pools/rfast/containers/mysql dev=rfast/containers/mysql
DBUG[06-08|12:13:37] MountInstance finished                   pool=rfast driver=zfs project=default instance=mysql
DBUG[06-08|12:13:37] UpdateInstanceBackupFile started         pool=rfast driver=zfs project=default instance=mysql
DBUG[06-08|12:13:37] UpdateInstanceBackupFile finished        pool=rfast driver=zfs project=default instance=mysql
DBUG[06-08|12:13:37] MountInstance started                    pool=rfast driver=zfs project=default instance=mysql
DBUG[06-08|12:13:37] MountInstance finished                   pool=rfast driver=zfs project=default instance=mysql
INFO[06-08|12:13:37] Starting container                       project=default name=mysql action=start created=2020-05-31T23:45:31-0500 ephemeral=false used=2020-06-08T11:01:39-0500 stateful=false
DBUG[06-08|12:13:37] Handling                                 method=GET url=/internal/containers/56/onstart ip=@ user=
DBUG[06-08|12:13:37] MountInstance started                    driver=zfs pool=rfast project=default instance=mysql
DBUG[06-08|12:13:37] MountInstance finished                   driver=zfs pool=rfast project=default instance=mysql
DBUG[06-08|12:13:37] Scheduler: container mysql started: re-balancing 
INFO[06-08|12:13:37] Started container                        stateful=false project=default name=mysql action=start created=2020-05-31T23:45:31-0500 ephemeral=false used=2020-06-08T11:01:39-0500
DBUG[06-08|12:13:37] Scheduler: network: veth99fe1ce7 has been added: updating network priorities 
DBUG[06-08|12:13:37] Scheduler: network: veth9cc2130c has been added: updating network priorities 
DBUG[06-08|12:13:37] MountInstance started                    driver=zfs pool=lxcpool project=default instance=nflfan
DBUG[06-08|12:13:37] Mounted ZFS dataset                      driver=zfs pool=lxcpool dev=rpool/lxc/containers/nflfan path=/var/snap/lxd/common/lxd/storage-pools/lxcpool/containers/nflfan
DBUG[06-08|12:13:37] MountInstance finished                   driver=zfs pool=lxcpool project=default instance=nflfan
DBUG[06-08|12:13:37] UpdateInstanceBackupFile started         driver=zfs pool=lxcpool project=default instance=nflfan
DBUG[06-08|12:13:38] UpdateInstanceBackupFile finished        driver=zfs pool=lxcpool project=default instance=nflfan
DBUG[06-08|12:13:38] MountInstance started                    driver=zfs pool=lxcpool project=default instance=nflfan
DBUG[06-08|12:13:38] MountInstance finished                   driver=zfs pool=lxcpool project=default instance=nflfan
INFO[06-08|12:13:38] Starting container                       ephemeral=false used=2020-06-08T11:01:39-0500 stateful=false project=default name=nflfan action=start created=2020-04-09T14:52:12-0500
DBUG[06-08|12:13:38] Handling                                 url=/internal/containers/53/onstart ip=@ user= method=GET
DBUG[06-08|12:13:38] MountInstance started                    driver=zfs pool=lxcpool project=default instance=nflfan
DBUG[06-08|12:13:38] MountInstance finished                   driver=zfs pool=lxcpool project=default instance=nflfan
DBUG[06-08|12:13:38] Scheduler: container nflfan started: re-balancing 
INFO[06-08|12:13:38] Started container                        stateful=false project=default name=nflfan action=start created=2020-04-09T14:52:12-0500 ephemeral=false used=2020-06-08T11:01:39-0500
DBUG[06-08|12:13:38] Scheduler: network: veth513b28d7 has been added: updating network priorities 
DBUG[06-08|12:13:38] Scheduler: network: vetha064dfec has been added: updating network priorities 
DBUG[06-08|12:13:38] MountInstance started                    driver=zfs pool=lxcpool instance=nginx project=default
DBUG[06-08|12:13:38] Mounted ZFS dataset                      driver=zfs pool=lxcpool dev=rpool/lxc/containers/nginx path=/var/snap/lxd/common/lxd/storage-pools/lxcpool/containers/nginx
DBUG[06-08|12:13:38] MountInstance finished                   driver=zfs pool=lxcpool instance=nginx project=default
DBUG[06-08|12:13:38] UpdateInstanceBackupFile started         driver=zfs pool=lxcpool project=default instance=nginx
DBUG[06-08|12:13:38] UpdateInstanceBackupFile finished        driver=zfs pool=lxcpool project=default instance=nginx
DBUG[06-08|12:13:38] MountInstance started                    driver=zfs pool=lxcpool project=default instance=nginx
DBUG[06-08|12:13:38] MountInstance finished                   driver=zfs pool=lxcpool project=default instance=nginx
INFO[06-08|12:13:38] Starting container                       ephemeral=false used=2020-06-08T11:01:40-0500 stateful=false project=default name=nginx action=start created=2020-04-09T14:57:01-0500
DBUG[06-08|12:13:38] Handling                                 method=GET url=/internal/containers/54/onstart ip=@ user=
DBUG[06-08|12:13:38] MountInstance started                    driver=zfs pool=lxcpool project=default instance=nginx
DBUG[06-08|12:13:38] MountInstance finished                   driver=zfs pool=lxcpool project=default instance=nginx
DBUG[06-08|12:13:38] Scheduler: container nginx started: re-balancing 
INFO[06-08|12:13:38] Started container                        used=2020-06-08T11:01:40-0500 stateful=false project=default name=nginx action=start created=2020-04-09T14:57:01-0500 ephemeral=false
DBUG[06-08|12:13:38] Scheduler: network: vethdb76a6fb has been added: updating network priorities 
DBUG[06-08|12:13:38] Scheduler: network: veth37c487e6 has been added: updating network priorities 
DBUG[06-08|12:13:38] MountInstance started                    driver=zfs pool=lxcpool project=default instance=postfix
DBUG[06-08|12:13:38] Mounted ZFS dataset                      driver=zfs pool=lxcpool dev=rpool/lxc/containers/postfix path=/var/snap/lxd/common/lxd/storage-pools/lxcpool/containers/postfix
DBUG[06-08|12:13:38] MountInstance finished                   driver=zfs pool=lxcpool project=default instance=postfix
DBUG[06-08|12:13:38] UpdateInstanceBackupFile started         driver=zfs pool=lxcpool project=default instance=postfix
DBUG[06-08|12:13:38] UpdateInstanceBackupFile finished        driver=zfs pool=lxcpool project=default instance=postfix
DBUG[06-08|12:13:38] MountInstance started                    driver=zfs pool=lxcpool project=default instance=postfix
DBUG[06-08|12:13:38] MountInstance finished                   driver=zfs pool=lxcpool project=default instance=postfix
INFO[06-08|12:13:38] Starting container                       project=default name=postfix action=start created=2020-04-08T23:12:14-0500 ephemeral=false used=2020-06-08T11:01:40-0500 stateful=false
DBUG[06-08|12:13:39] Handling                                 method=GET url=/internal/containers/52/onstart ip=@ user=
DBUG[06-08|12:13:39] MountInstance started                    driver=zfs pool=lxcpool project=default instance=postfix
DBUG[06-08|12:13:39] MountInstance finished                   driver=zfs pool=lxcpool project=default instance=postfix
DBUG[06-08|12:13:39] Scheduler: container postfix started: re-balancing 
INFO[06-08|12:13:39] Started container                        project=default name=postfix action=start created=2020-04-08T23:12:14-0500 ephemeral=false used=2020-06-08T11:01:40-0500 stateful=false
DBUG[06-08|12:13:39] Scheduler: network: vethe4522a89 has been added: updating network priorities 
DBUG[06-08|12:13:39] Scheduler: network: vethdf45e64b has been added: updating network priorities 
DBUG[06-08|12:13:39] MountInstance started                    driver=dir pool=lxd project=default instance=tinydns-net
DBUG[06-08|12:13:39] MountInstance finished                   driver=dir pool=lxd project=default instance=tinydns-net
DBUG[06-08|12:13:39] UpdateInstanceBackupFile started         driver=dir pool=lxd project=default instance=tinydns-net
DBUG[06-08|12:13:39] UpdateInstanceBackupFile finished        driver=dir pool=lxd project=default instance=tinydns-net
DBUG[06-08|12:13:39] MountInstance started                    driver=dir pool=lxd project=default instance=tinydns-net
DBUG[06-08|12:13:39] MountInstance finished                   driver=dir pool=lxd project=default instance=tinydns-net
INFO[06-08|12:13:39] Starting container                       used=2020-06-08T11:01:41-0500 stateful=false project=default name=tinydns-net action=start created=2017-04-13T12:29:00-0500 ephemeral=false
DBUG[06-08|12:13:39] Handling                                 url=/internal/containers/7/onstart ip=@ user= method=GET
DBUG[06-08|12:13:39] MountInstance started                    driver=dir pool=lxd project=default instance=tinydns-net
DBUG[06-08|12:13:39] MountInstance finished                   driver=dir pool=lxd project=default instance=tinydns-net
DBUG[06-08|12:13:39] Scheduler: container tinydns-net started: re-balancing 
INFO[06-08|12:13:40] Started container                        name=tinydns-net action=start created=2017-04-13T12:29:00-0500 ephemeral=false used=2020-06-08T11:01:41-0500 stateful=false project=default
DBUG[06-08|12:13:40] Success for task operation: 7b4bf1a8-461f-4310-b1e8-c7a73e83063d 
DBUG[06-08|12:13:40] Image already exists in the db           image=d125be3eded77d54fdeb2971e25b7f98b8e80a71f365353def505ac0c7c63803
DBUG[06-08|12:13:40] Image already exists on storage pool "rfast" 
DBUG[06-08|12:13:40] Already up to date                       fp=d125be3eded77d54fdeb2971e25b7f98b8e80a71f365353def505ac0c7c63803
DBUG[06-08|12:13:40] Success for task operation: 5f4b38f0-1a3a-4982-a87d-1115adda6c4c

Thanks in advance for your help!

**Comment by stgraber** [comment]
Can you show `ps fauxww` and `snap changes`?

**Comment by bakerjalexander** [comment]
snap changes:

ID   Status  Spawn                   Ready                   Summary
268  Error   yesterday at 17:11 CDT  yesterday at 17:32 CDT  Auto-refresh snap "lxd"
269  Error   yesterday at 20:51 CDT  yesterday at 21:12 CDT  Auto-refresh snap "lxd"
270  Error   today at 01:10 CDT      today at 01:31 CDT      Refresh "lxd" snap
271  Error   today at 01:34 CDT      today at 01:54 CDT      Remove "lxd" snap
272  Error   today at 03:00 CDT      today at 03:20 CDT      Refresh "lxd" snap
273  Error   today at 03:56 CDT      today at 04:17 CDT      Auto-refresh snap "lxd"
274  Error   today at 07:41 CDT      today at 08:01 CDT      Auto-refresh snap "lxd"
275  Error   today at 10:02 CDT      today at 10:23 CDT      Refresh "lxd" snap
276  Done    today at 10:30 CDT      today at 10:30 CDT      Running service command
277  Error   today at 10:34 CDT      today at 11:07 CDT      Refresh "lxd" snap
278  Error   today at 11:07 CDT      today at 11:28 CDT      Install "lxd" snap from file "lxd_15359.snap"
279  Error   today at 11:34 CDT      today at 11:54 CDT      Refresh "lxd" snap
280  Done    today at 11:55 CDT      today at 11:55 CDT      Running service command
281  Done    today at 12:05 CDT      today at 12:05 CDT      Running service command
282  Error   today at 12:14 CDT      today at 12:34 CDT      Refresh "lxd" snap
283  Error   today at 13:07 CDT      today at 13:27 CDT      Auto-refresh snap "lxd"
284  Error   today at 16:27 CDT      today at 16:27 CDT      Running service command
285  Done    today at 16:28 CDT      today at 16:28 CDT      Running service command
286  Done    today at 16:30 CDT      today at 16:30 CDT      Running service command

[ps.fauxww.txt](https://github.com/lxc/lxd/files/4748747/ps.fauxww.txt)

**Comment by stgraber** [comment]
So one thing that may not be helping is that you have a side-loaded LXD binary at `/var/snap/lxd/common/lxd.debug`.

You probably should remove that, then do `systemctl reload snap.lxd.daemon`, check `ps fauxww` again to confirm that LXD did get restarted and isn't using `lxd.debug` anymore, then attempt the refresh again.

**Comment by bakerjalexander** [comment]
Did
rm /var/snap/lxd/common/lxd.debug
systemctl reload snap.lxd.daemon
check with ps and lxd shows running as 
lxd --logfile /var/snap/lxd/common/lxd/logs/lxd.log --group lxd --debug
(I had set debug to true from before)
same result as before:  
INFO Waiting for snap.lxd.activate.service to stop.
2020-06-08T21:23:54-05:00 INFO snap.lxd.activate.service refused to stop, killing.
2020-06-08T21:24:00-05:00 INFO Waiting for snap.lxd.daemon.unix.socket to stop
last command repeats 20 times, etc., same as before

Wondering if there is a problem with the systemd files.  Here's what I've got in /etc/systemd/system:

snap-core18-1705.mount            snap-lxd-14890.mount     snap-lxd-15272.mount
snap-core18-1754.mount            snap-lxd-14954.mount     snap-lxd-15359.mount
snap-core-9066.mount                snap-lxd-15067.mount     snap.lxd.activate.service
snap-core-9289.mount                snap-lxd-15138.mount     snap.lxd.daemon.service
'snap-hello\x2dworld-27.mount'   snap-lxd-15161.mount     snap.lxd.daemon.unix.socket
'snap-hello\x2dworld-29.mount'   snap-lxd-15223.mount

Last three all point to Requires=snap-lxd-14954.mount

What signals a successful exit to "INFO Waiting for snap.lxd.daemon.unix.socket to stop" because the socket file is still there:

srw-rw----  1 root lxd      0 Jun  8 21:20 unix.socket

**Comment by stgraber** [comment]
Oh, hold on, the error is about `snap.lxd.active`, not about the LXD daemon...

**Comment by bakerjalexander** [comment]
This morning the lxd daemon had stopped again, I assume as the result of an auto-refresh.  I had to restart the daemon with:

systemctl start snap.lxd.activate.service 

and it answered back with:

Job for snap.lxd.activate.service canceled.

But the daemon did start at least. This is from the last log from the night before:

t=2020-06-09T00:17:37-0500 lvl=info msg="Received 'terminated signal', exiting" 
t=2020-06-09T00:17:37-0500 lvl=info msg="Starting shutdown sequence" 
t=2020-06-09T00:17:37-0500 lvl=info msg="Stopping REST API handler:" 
t=2020-06-09T00:17:37-0500 lvl=info msg=" - closing socket" socket=/var/snap/lxd/common/lxd/unix.socket
t=2020-06-09T00:17:37-0500 lvl=info msg="Stopping /dev/lxd handler:" 
t=2020-06-09T00:17:37-0500 lvl=info msg=" - closing socket" socket=/var/snap/lxd/common/lxd/devlxd/sock
t=2020-06-09T00:17:37-0500 lvl=info msg="Closing the database" 
t=2020-06-09T00:17:37-0500 lvl=dbug msg="Stop database gateway" 
t=2020-06-09T00:17:38-0500 lvl=dbug msg="Not unmounting temporary filesystems (containers are still running)"

**Comment by stgraber** [comment]
Can you show `systemctl -a`?

**Comment by stgraber** [comment]
The activate unit should only start once per system boot and should not be started manually, so I'm a bit confused as to why it seems to be holding the refresh.

**Comment by bakerjalexander** [comment]
My only thought is some kind of failure to note the destruction of the socket file. Does it have something to do with the containers still running? I can see that it is not being destroyed while snap.lxd.daemon.unix.socket is trying to stop.
[systemctl.a.txt](https://github.com/lxc/lxd/files/4752557/systemctl.a.txt)

**Comment by bakerjalexander** [comment]
Seems like I could just destroy the socket file manually and somehow have the systemd file stop command read as /bin/true but not sure exactly how to do that. I tried it and it didn't work but again, not sure what I'm doing there.

**Comment by stgraber** [comment]
Ah yeah, that system isn't in a great shape here... Normally `lxd.activate` should be inactive/exited, `unix.socket` should always be active and `service` should be active if LXD is running.

Can you try:
 - echo "reload" > /var/snap/lxd/common/state
 - systemctl stop snap.lxd.daemon
 - systemctl start snap.lxd.daemon.unix.socket
 - systemctl start snap.lxd.activate

**Comment by stgraber** [comment]
That should let you stop LXD without stopping any of the containers, then getting the unix socket back to working order and finally running activate as it would on startup which should then start LXD and get all 3 units in the desired state.

**Comment by bakerjalexander** [comment]
Did
    echo "reload" > /var/snap/lxd/common/state
    systemctl stop snap.lxd.daemon
    systemctl start snap.lxd.daemon.unix.socket
    systemctl start snap.lxd.activate

Answer:

Job for snap.lxd.activate.service canceled.

Attached is new sysctl -a after the above commands
[systemctl.a.new.txt](https://github.com/lxc/lxd/files/4752643/systemctl.a.new.txt)

**Comment by stgraber** [comment]
Ok, systemd looks quite confused by that unit...

Can you show `systemctl status snap.lxd.activate.service`, `journalctl -u snap.lxd.activate.service -n 30` and `systemctl cat snap.lxd.activate.service`?

**Comment by bakerjalexander** [comment]
Attached.  BTW, I had changed the TimeoutStopSec=10 on the snap.lxd.activate.service file.
[cat.snap.lxd.activate.service.txt](https://github.com/lxc/lxd/files/4752698/cat.snap.lxd.activate.service.txt)
[journalctl.txt](https://github.com/lxc/lxd/files/4752699/journalctl.txt)
[status.snap.lxd.activate.service.txt](https://github.com/lxc/lxd/files/4752700/status.snap.lxd.activate.service.txt)

**Comment by stgraber** [comment]
Are any of PID 1170 or 1087 still unning?

**Comment by stgraber** [comment]
*running

**Comment by bakerjalexander** [comment]
Neither one is running. "lxc list" gives
Error: Get "http://unix.socket/1.0": dial unix /var/snap/lxd/common/lxd/unix.socket: connect: connection refused
but the unix.socket file is still there

looks like it went down about 90 minutes ago

t=2020-06-09T11:22:40-0500 lvl=info msg="Received 'terminated signal', exiting" 
t=2020-06-09T11:22:40-0500 lvl=info msg="Starting shutdown sequence" 
t=2020-06-09T11:22:40-0500 lvl=info msg="Stopping REST API handler:" 
t=2020-06-09T11:22:40-0500 lvl=info msg=" - closing socket" socket=/var/snap/lxd/common/lxd/unix.socket
t=2020-06-09T11:22:40-0500 lvl=info msg="Stopping /dev/lxd handler:" 
t=2020-06-09T11:22:40-0500 lvl=info msg=" - closing socket" socket=/var/snap/lxd/common/lxd/devlxd/sock
t=2020-06-09T11:22:40-0500 lvl=info msg="Closing the database" 
t=2020-06-09T11:22:40-0500 lvl=dbug msg="Stop database gateway" 
t=2020-06-09T11:22:40-0500 lvl=dbug msg="Not unmounting temporary filesystems (containers are still running)"

**Comment by stgraber** [comment]
Can you show:
 - `systemctl -a | grep lxd`
 - `netstat -lnp | grep unix.socket`

**Comment by bakerjalexander** [comment]
OK, this is without restarting the daemon:
netstat -lnp | grep unix.socket is null, nothing returned

systemctl file is attached
[systemctl.noreload.txt](https://github.com/lxc/lxd/files/4754455/systemctl.noreload.txt)

**Comment by bakerjalexander** [comment]
Here's an ls of /var/snap/lxd

drwxr-xr-x 2 root root  2 Jun 12  2018 14954
drwxr-xr-x 2 root root  2 Jun 12  2018 14890
lrwxrwxrwx 1 root root  5 May  5 21:57 current -> 14954
drwxr-xr-x 7 root root 12 Jun  9 09:10 common

**Comment by stgraber** [comment]
Ok, can you confirm that there is no LXD daemon running according to `ps fauxww`?

If there isn't, then in one terminal try running: `lxd --debug --group lxd` which should start LXD and then just hang as it's running.
With that running, in another terminal run `lxd activateifneeded --debug` to see if they can communicate and if there's anything weird going on there.

**Comment by bakerjalexander** [comment]
OK, no lxd was running, started as per instruction. Ran "lxd activateifneeded --debug" got

DBUG[06-09|15:12:45] Daemon has auto-started instances, activating... 
DBUG[06-09|15:12:45] Connecting to a local LXD over a Unix socket 
DBUG[06-09|15:12:45] Sending request to LXD method=GET url=http://unix.socket/1.0 etag=
DBUG[06-09|15:12:45] Got response struct from LXD 

and then a json dump of the config file

**Comment by stgraber** [comment]
Ok, so that looks correct.
Try:
 - Kill that manually started debug LXD daemon
 - systemctl start snap.lxd.daemon.unix.socket
 - lxc info

This should enable socket activation and then have `lxc info` trigger it, causing the daemon to start before it shows you the daemon info.

**Comment by bakerjalexander** [comment]
OK, got the text version of the config file. Shall I try "snap refresh lxd" ?

**Comment by bakerjalexander** [comment]
Well I went ahead and tried "snap refresh lxd" and still got the same results as my first post.

**Comment by stgraber** [comment]
Looks like there's something really wrong with systemd...
If it's an option, a reboot may clear up whatever broken state there is.

If not, you may want to try `systemctl mask snap.lxd.activate.service` to get rid of that service which may or may not help snapd. But you'll need to remember to unmask it or LXD won't start on reboot.

**Comment by bakerjalexander** [comment]
Maybe this is some help:

If I run "snap stop lxd", lxd will exit, but if I try "snap start lxd" I get :

error: cannot perform the following tasks:
- start of [lxd.activate lxd.daemon] (# systemctl start snap.lxd.activate.service snap.lxd.daemon.service

and them some messages like 

Got message type=signal sender=org.freedesktop.systemd1 destination=n/a path=/org/freedesktop/systemd1 interface=org.freedesktop.systemd1.Manager member=JobRemoved cookie=12 reply_cookie=0 signature=uoss error-name=n/a error-message=n/a
Got result done/Resource temporarily unavailable for job snap.lxd.daemon.service
...more messages of the same with freedesktop.etc,etc

and then
Job for snap.lxd.activate.service canceled.
Got result canceled/Operation canceled for job snap.lxd.activate.service
Bus n/a: changing state RUNNING → CLOSED
)
- start of [lxd.activate lxd.daemon] (exit status 1)

but the lxd daemon runs.
Also, this process remains:
/bin/sh /snap/lxd/14954/commands/daemon.start
and killing it just causes it to respawn

Any ideas?

**Comment by stgraber** [comment]
Yeah, would need root access to poke at LXD and see what's up with it.

```
ssh-ed25519 AAAAC3NzaC1lZDI1NTE5AAAAIOhdeiF51bz5W+mPtpRYnXQRuLqimMPE5PBD1ab8APt6 cardno:000611576019
```

**Comment by bakerjalexander** [comment]
I'm OK with doing that but going to try one last thing. For whatever reason, I've got some apt sources for the old lxcfs daemon, et al running:
liblxc-common                3.0.3-0ubuntu1~18.04.1
liblxc1                             3.0.3-0ubuntu1~18.04.1                            
lxc                                   3.0.3-0ubuntu1~18.04.1                            
lxc-common                    2.1.1-0ubuntu1~ubuntu16.04.1~ppa1                 
lxc-desktop                     0.1                                               
lxc-templates                  3.0.0-0ubuntu1                                    
lxc-utils                           3.0.3-0ubuntu1~18.04.1                            
lxc1                                 3.0.0-0ubuntu2~ubuntu16.04.1~ppa1                 

Do you think this might have anything to do with my problems? If I purge them, will the snap version of those packages takeover? I've got a few containers left running under the old system but fairly confident I can move them over.

Thanks
lxcfs                                  3.0.3-0ubuntu1~18.04.2                            
python3-lxc                            3.0.1-0ubuntu1

**Comment by stgraber** [comment]
No, having those packages would be harmless. You can certainly purge them as they would be completely unused right now, but they also can't really conflict with anything.

**Comment by bakerjalexander** [comment]
I'd like to try one last thing: manually removing all traces of snap and its elements and then reloading.  Tell me if this is right:
backup /var/snap/lxd/common/lxd
backup lxd config file
stop all lxc and lxd processes
apt purge snap
remove /var/snap directory
remove all systemd snap files
And then
Reinstall snap and lxd
reinitialize lxd from backup config file
stop snap, lxd
restore backup /var/snap/lxd/common/lxd
restart snap, lxd
cross fingers

Am I missing anything?

Thanks

**Comment by stgraber** [comment]
Yeah, that should work assuming you're not using something too weird as a storage backend.

**Comment by bakerjalexander** [comment]
I manually removed all traces of snap from my system, redid all of my containers and all was going well until 
snapd[9206]: storehelpers.go:438: cannot refresh: snap has no updates available: "core18", "snapd
"
And now I'm back to the same problem.

I had systemd debugging on so I'm including about 80 seconds of syslog from that point. It appears that something is triggering systemd to reload but it can't find the files and lxd is never restarted

lxc info gives 

Error: Get "http://unix.socket/1.0": dial unix /var/snap/lxd/common/lxd/unix.socket: connect: connection refused

and snap refresh lxd gives repeated

2020-06-15T21:28:36-05:00 INFO Waiting for snap.lxd.daemon.unix.socket to stop.

and then repeated

2020-06-15T21:38:36-05:00 INFO Waiting for snap.lxd.daemon.service to stop.

Ending with

error: cannot perform the following tasks:
- Mount snap "lxd" (15564) (snap-lxd-15564.mount failed to stop: timeout)
- Stop snap "lxd" services (snap.lxd.daemon.unix.socket failed to stop: timeout)

I can restart the service with service snap.lxd.daemon start

I'm attaching 80 seconds worth of the syslog so you can see what systemd is doing.

Thanks for any advice!
[syslog.80sec.txt](https://github.com/lxc/lxd/files/4784209/syslog.80sec.txt)

**Comment by stgraber** [comment]
We're going to need to look at such a system directly, clearly none of the usual logs and stuff we looked at are particularly useful and you're the only one who has ever reported this so it's not something we can reproduce on our side.

**Comment by stgraber** [comment]
@bakerjalexander any update on this?

**Comment by stgraber** [comment]
Closing for inactivity. I've had a colleague run into something very very similar where snapd was stating "Waiting for XYZ to stop" yet all units were in the correct stopped state already. This suggests some kind of bug in snapd.

We'll have the snapd team investigate that colleague's system so they can hopefully sort this out.

If anyone else hits this, feel free to comment here so we can get an idea of the scale of the problem.
Also if you found a way around it, please let us know :)

**Comment by bakerjalexander** [comment]
Thanks. Yeah I gave up on this.  Apparently the refresh issue is a problem
for many.  I found a post on how to stop the thing from refreshing on its
own and I'll pass it along for anybody that stumbles on this post.

Add
1 1 1 * * /usr/bin/snap set system refresh.hold="2030-01-01T01:00:00-01:00"
To your crontab

I'm going to work on a script to manually download and install the newest
version of lxd without using snap refresh.  It appears that the only thing
that need happen is to update a systemd file, rsync everything but the
config/db files to /var/snap/lxd and change a symlink.

I really wish snap could just duplicate "apt purge" or "apt install
--reinstall." When I did snap remove myself I ran into the same issue with
the waiting on the socket to close thing.

Snap should be very useful to lxd and linux in general, but it's definitely
got issues.

Anyways, I appreciate the time you put in on this and I'm sorry it was
fruitless.

John

**Comment by cjohnston1158** [comment]
After working with the snappy team, the issue I was seeing was found that it's due to having debug in the kernel command line.

https://bugs.launchpad.net/snapd/+bug/1885597

**Comment by bakerjalexander** [comment]
Can confirm. Removed debug from the kernel command line and all is well!

Thanks!

**Comment by crass** [comment]
For anyone looking for a work around that allows keeping `debug` as a kernel parameter see [this comment](https://bugs.launchpad.net/snapd/+bug/1885597/comments/3) on the above mentioned bug report. The key is to use `systemd.log_level=info` on the commandline to override systemd interpretation of the `debug` parameter.
